castle/cms/subscribers.py

```diff
@@ -209,3 +217,111 @@ def on_youtube_video_state_changed(obj, event):
             tasks.youtube_video_state_changed.delay(obj)
         except CannotGetPortalError:
             pass
+
+
+def create_version_on_save(context, event):
+    """Creates a new version on a versionable object when the object is saved.
+    A new version is created if the type is automatic versionable and has
+    changed or if the user has entered a change note.
+    """
+    # according to Products.CMFEditions' update_version_on_edit script
+
+    # only version the modified object, not its container on modification
+    if IContainerModifiedEvent.providedBy(event):
+        return
+
+    # XXX dirty hack for stagingbehavior, which triggers a event with
+    # a aq_based context when deleting the working copy
+    try:
+        pr = context.portal_repository
+    except AttributeError:
+        return
+
+    if not pr.isVersionable(context):
+        # cancel, the object is not versionable
+        return
+
+    create_version = False
+
+    if getattr(context, 'REQUEST', None):
+        changeNote = get_change_note(context.REQUEST, None)
+    else:
+        changeNote = None
+
+    if changeNote:
+        # user has entered a change note. create a new version even if nothing
+        # has changed.
+        create_version = True
+
+    elif pr.supportsPolicy(context, 'at_edit_autoversion'):
+        # automatic versioning is enabled for this portal type
+
+        if not base_hasattr(context, 'version_id'):
+            # we do not have a initial version
+            create_version = True
+        else:
+            try:
+                create_version = not pr.isUpToDate(context, context.version_id)
+            except ArchivistUnregisteredError:
+                # The object is not actually registered, but a version is
+                # set, perhaps it was imported, or versioning info was
+                # inappropriately destroyed
+                create_version = True
+
+    # create new version if needed
+    if create_version:
+        pr.save(obj=context, comment=changeNote)
+
+
+def create_initial_version_after_adding(context, event):
```

Comment:
was this pulled from some where? if it was, it'd be nice to have a reference to the original in a comment